File: backend/models/rotation_rule.py
"""
Rotation Rule model for traffic scheduling
"""

from sqlalchemy import Column, Integer, String, DateTime, Boolean, ForeignKey, Text, CheckConstraint, text
from sqlalchemy.dialects.postgresql import UUID, JSONB
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
from backend.database import Base
import enum


class RotationType(str, enum.Enum):
    """Rotation type enumeration"""
    SEQUENTIAL = "SEQUENTIAL"  # Play in order
    RANDOM = "RANDOM"  # Random order
    WEIGHTED = "WEIGHTED"  # Weighted random
    EVEN = "EVEN"  # Even distribution


class RotationRule(Base):
    """Rotation Rule model for managing spot rotation"""
    __tablename__ = "rotation_rules"

    id = Column(UUID(as_uuid=True), primary_key=True, server_default=text("gen_random_uuid()"), index=True)
    name = Column(String(100), nullable=False, index=True)
    description = Column(Text, nullable=True)
    
    # Rotation configuration
    rotation_type = Column(String(20), nullable=False, default=RotationType.SEQUENTIAL.value)
    daypart_id = Column(UUID(as_uuid=True), ForeignKey("dayparts.id"), nullable=True, index=True)
    campaign_id = Column(UUID(as_uuid=True), ForeignKey("campaigns.id"), nullable=True, index=True)
    
    # Rotation parameters
    min_separation = Column(Integer, default=0)  # Minimum spots between same campaign
    max_per_hour = Column(Integer, nullable=True)  # Maximum spots per hour
    max_per_day = Column(Integer, nullable=True)  # Maximum spots per day
    weights = Column(JSONB, nullable=True)  # Weighted rotation weights {campaign_id: weight}
    
    # Cut-specific rotation support
    cut_specific = Column(Boolean, default=False)  # Whether this rule applies to specific cuts
    cut_weights = Column(JSONB, nullable=True)  # Weighted rotation weights {cut_id: weight}
    program_specific = Column(Boolean, default=False)  # Whether this rule is program-specific
    
    # Scheduling rules
    exclude_days = Column(JSONB, nullable=True)  # Days to exclude (0=Monday, 6=Sunday)
    exclude_times = Column(JSONB, nullable=True)  # Time ranges to exclude
    priority = Column(Integer, default=0)  # Rule priority (higher = more important)
    
    # Status
    active = Column(Boolean, default=True, index=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
    
    # Relationships
    daypart = relationship("Daypart", foreign_keys=[daypart_id])
    campaign = relationship("Campaign", foreign_keys=[campaign_id])

    def __repr__(self):
        return f"<RotationRule(name='{self.name}', type='{self.rotation_type}')>"

File: archive/librelog-web-api-original/src/main/java/com/onelpro/librelog/models/RotationRule.java
package com.onelpro.librelog.models;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Index;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.Instant;
import java.util.UUID;

/**
 * Entity representing a rotation rule in the system.
 */
@Entity
@Table(name = "rotation_rules", indexes = {
        @Index(name = "ix_rotation_rules_id", columnList = "id"),
        @Index(name = "ix_rotation_rules_name", columnList = "name"),
        @Index(name = "ix_rotation_rules_daypart_id", columnList = "daypart_id"),
        @Index(name = "ix_rotation_rules_campaign_id", columnList = "campaign_id"),
        @Index(name = "ix_rotation_rules_active", columnList = "active")
})
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class RotationRule {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(name = "name", nullable = false, length = 100)
    private String name;

    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    @Column(name = "rotation_type", nullable = false, length = 20)
    @Builder.Default
    private String rotationType = "SEQUENTIAL";

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "daypart_id")
    private Daypart daypart;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "campaign_id")
    private Campaign campaign;

    @Column(name = "min_separation")
    @Builder.Default
    private Integer minSeparation = 0;

    @Column(name = "max_per_hour")
    private Integer maxPerHour;

    @Column(name = "max_per_day")
    private Integer maxPerDay;

    @Column(name = "weights", columnDefinition = "JSONB")
    private String weights;

    @Column(name = "exclude_days", columnDefinition = "JSONB")
    private String excludeDays;

    @Column(name = "exclude_times", columnDefinition = "JSONB")
    private String excludeTimes;

    @Column(name = "priority")
    @Builder.Default
    private Integer priority = 0;

    @Column(name = "active")
    @Builder.Default
    private Boolean active = true;

    @Column(name = "created_at", nullable = false)
    private Instant createdAt;

    @Column(name = "updated_at", nullable = false)
    private Instant updatedAt;
}

